File: src/test/kotlin/official/kirizu/kirivers/client/AdapterTest.kt
package official.kirizu.kirivers.client

import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.jsonObject
import java.nio.file.Files
import java.text.Normalizer
import java.util.zip.ZipEntry
import java.util.zip.ZipOutputStream
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertIs
import kotlin.test.assertTrue

class AdapterTest {
    @Test
    fun nfcAndSlashNormalizationMatchesServerRules() {
        val nfd = Normalizer.normalize("é", Normalizer.Form.NFD)
        val nfc = Normalizer.normalize("é", Normalizer.Form.NFC)
        assertEquals(InstallPaths.normalize("dir\\café"), InstallPaths.normalize("dir/café"))
        assertEquals(InstallPaths.normalize("a/${nfd}/b"), InstallPaths.normalize("a/${nfc}/b"))
        assertFailsWith<InvalidPathException> { InstallPaths.normalize("../x") }
        assertFailsWith<InvalidPathException> { InstallPaths.normalize("/abs") }
        assertFailsWith<InvalidPathException> { InstallPaths.normalize("C:windows") }
    }

    @Test
    fun updaterCapabilitiesFollowAdapters() {
        val transport = RecordingTransport { jsonOk(CHECK_200) }
        val client = Client(ClientConfig("http://example.test", "p", transport = transport))
        val none = Updater(client, fileStore = null, archiveUnpacker = null, patcher = null)
        assertEquals(listOf(Capability.FULL_PACKAGE) to null, none.capabilities())

        val withZip = Updater(client, fileStore = null, archiveUnpacker = JdkZipUnpacker(), patcher = null)
        assertEquals(
            listOf(Capability.FULL_PACKAGE, Capability.PATCH_PACKAGE) to null,
            withZip.capabilities(),
        )

        val tmp = Files.createTempDirectory("kv-fs")
        val store = JdkFileStore(tmp)
        val withFiles = Updater(client, fileStore = store, archiveUnpacker = null, patcher = null)
        assertTrue(withFiles.capabilities().first.contains(Capability.FILE_LIST))

        val patcher = object : Patcher {
            override fun supportedAlgos() = listOf(DeltaAlgo.BSDIFF)
            override fun apply(oldBytes: ByteArray, delta: ByteArray, algo: String) = oldBytes
        }
        val withPatch = Updater(client, fileStore = null, archiveUnpacker = null, patcher = patcher)
        val (caps, algos) = withPatch.capabilities()
        assertTrue(caps.contains(Capability.BINARY_DELTA))
        assertEquals(listOf(DeltaAlgo.BSDIFF), algos)

        client.check(
            CheckRequest(
                currentVersion = "1.0.0",
                os = "windows",
                arch = "x86_64",
                capabilities = caps,
                acceptedDeltaAlgos = algos,
            ),
        )
        val body = KiriversJson.parseToJsonElement(transport.last().body!!.decodeToString()).jsonObject
        assertTrue(body["capabilities"].toString().contains("binary_delta"))
        assertTrue(body["accepted_delta_algos"].toString().contains("bsdiff"))
    }

    @Test
    fun packPollRepeatsIdenticalJson() {
        var hits = 0
        val transport = RecordingTransport { req ->
            if (RecordingTransport.pathOf(req.url).endsWith("/update/pack")) {
                hits += 1
                if (hits < 3) jsonOk("""{"status":"pending"}""", 202)
                else jsonOk("""{"status":"ready","package_url":"/pkg","sha256":"00"}""")
            } else {
                errorJson(404, "NOT_FOUND", "no")
            }
        }
        val client = Client(ClientConfig("http://example.test", "p", transport = transport))
        val req = PackRequest("1.0.0", "1.1.0", "windows", "x86_64", neededPaths = listOf("a/b", "c"))
        val result = client.packUntilReady(req, PackPollOptions(initialDelayMs = 1, maxDelayMs = 2, deadlineMs = 5_000))
        assertEquals("ready", result.status)
        val bodies = transport.requests
            .filter { it.method == "POST" }
            .map { it.body!!.decodeToString() }
        assertEquals(3, bodies.size)
        assertEquals(bodies[0], bodies[1])
        assertEquals(bodies[1], bodies[2])
        assertTrue(bodies[0].contains("needed_paths"))
    }

    @Test
    fun unknownMagicDoesNotCrossDecodeAndUpdaterFallsBack() {
        assertFailsWith<UnknownDeltaMagicException> {
            DeltaMagic.require(DeltaAlgo.BSDIFF, "HDIFF13&xxxx".toByteArray())
        }
        assertTrue(DeltaMagic.matches(DeltaAlgo.XDELTA3, byteArrayOf(0xD6.toByte(), 0xC3.toByte(), 0xC4.toByte(), 1)))
        assertFalse(DeltaMagic.matches(DeltaAlgo.HDIFFPATCH, "BSDIFF40".toByteArray()))

        val full = byteArrayOf(9, 9, 9)
        val hasher = JdkHasher()
        val fullHash = hasher.sha256Hex(full)
        val check = CHECK_200.replace("\"delta_available\": false", "\"delta_available\": true")
            .replace("\"sha256\": \"aaaa\"", "\"sha256\": \"$fullHash\"")
        val transport = RecordingTransport { req ->
            when {
                RecordingTransport.pathOf(req.url).endsWith("/update/check") -> jsonOk(check)
                RecordingTransport.pathOf(req.url).endsWith("/update/diff") -> jsonOk(
                    """
                    {
                      "diff_mode": "binary_delta",
                      "root_hash": "",
                      "version_integer": null,
                      "version_semver": "1.1.0",
                      "channel": "stable",
                      "compare_engine": "semver",
                      "delta_algo": "bsdiff",
                      "package_url": "/delta",
                      "sha256": "dead"
                    }
                    """.trimIndent(),
                )
                RecordingTransport.pathOf(req.url).endsWith("/delta") ->
                    HttpResponse(200, body = "HDIFF13&nope".toByteArray())
                RecordingTransport.pathOf(req.url).contains("/packages/") ||
                    req.url.contains("aaaa") || req.url.endsWith("/abc") ->
                    HttpResponse(200, body = full)
                RecordingTransport.pathOf(req.url).endsWith("/telemetry/report") -> jsonOk("""{"status":"ok"}""", 202)
                else -> errorJson(404, "NOT_FOUND", RecordingTransport.pathOf(req.url))
            }
        }
        val client = Client(ClientConfig("http://example.test", "demo", transport = transport))
        val patcher = object : Patcher {
            override fun supportedAlgos() = listOf(DeltaAlgo.BSDIFF)
            override fun apply(oldBytes: ByteArray, delta: ByteArray, algo: String): ByteArray {
                DeltaMagic.require(algo, delta)
                return oldBytes
            }
        }
        val outcome = Updater(client, archiveUnpacker = null, patcher = patcher).run(
            UpdateIdentity("1.0.0", "windows", "x86_64", localPackage = byteArrayOf(1, 2, 3)),
        )
        assertTrue(outcome.fallbackToFull)
        assertEquals("full_package", outcome.diffMode)
        assertTrue(outcome.bytes.contentEquals(full))
    }

    @Test
    fun zipUnpackerMapsHashNamedMembers() {
        val hasher = JdkHasher()
        val payload = "hello-zip".toByteArray()
        val hash = hasher.sha256Hex(payload)
        val zipBytes = java.io.ByteArrayOutputStream().use { out ->
            ZipOutputStream(out).use { zip ->
                zip.putNextEntry(ZipEntry(hash))
                zip.write(payload)
                zip.closeEntry()
            }
            out.toByteArray()
        }
        val tmp = Files.createTempDirectory("kv-zip")
        val store = JdkFileStore(tmp)
        JdkZipUnpacker().unpack(
            zipBytes,
            listOf(
                IntegrityFile(
                    path = "bin/app.exe",
                    size = payload.size.toLong(),
                    installPolicy = "OVERWRITE",
                    sha256 = hash,
                ),
            ),
            store,
        )
        assertTrue(payload.contentEquals(store.read("bin/app.exe")))
    }

    @Test
    fun keepIfExistsOmittedFromNeededPaths() {
        val tmp = Files.createTempDirectory("kv-keep")
        val store = JdkFileStore(tmp)
        val bytes = "keep-me".toByteArray()
        store.write("data/local.dat", bytes)
        val hasher = JdkHasher()
        val transport = RecordingTransport { jsonOk(CHECK_200) }
        val client = Client(ClientConfig("http://example.test", "p", transport = transport))
        val updater = Updater(client, fileStore = store, archiveUnpacker = null)
        val needed = updater.neededPaths(
            IntegrityManifest(
                channel = "stable",
                packageType = "multi_file",
                files = listOf(
                    IntegrityFile(
                        path = "data/local.dat",
                        size = bytes.size.toLong(),
                        installPolicy = "KEEP_IF_EXISTS",
                        integrityCheck = true,
                        sha256 = hasher.sha256Hex(bytes),
                    ),
                    IntegrityFile(
                        path = "bin/app",
                        size = 1,
                        installPolicy = "OVERWRITE",
                        sha256 = "ff",
                    ),
                ),
            ),
            store,
        )
        assertEquals(listOf("bin/app"), needed)
    }

    @Test
    fun updaterWithoutReplacerStagesInsteadOfInstalling() {
        val payload = byteArrayOf(1, 2, 3, 4)
        val hash = JdkHasher().sha256Hex(payload)
        val check = CHECK_200.replace("\"sha256\": \"aaaa\"", "\"sha256\": \"$hash\"")
            .replace("\"size\": 4", "\"size\": ${payload.size}")
        val transport = RecordingTransport { req ->
            when {
                RecordingTransport.pathOf(req.url).endsWith("/update/check") -> jsonOk(check)
                RecordingTransport.pathOf(req.url).endsWith("/telemetry/report") -> jsonOk("{}", 202)
                else -> HttpResponse(200, body = payload)
            }
        }
        val client = Client(ClientConfig("http://example.test", "demo", transport = transport))
        val outcome = Updater(client, archiveUnpacker = null, replacer = null).run(
            UpdateIdentity("1.0.0", "windows", "x86_64"),
        )
        assertIs<UpdateKind>(outcome.kind)
        assertEquals(UpdateKind.STAGED, outcome.kind)
        assertFalse(outcome.applied)
        assertTrue(payload.contentEquals(outcome.bytes))
    }

    @Test
    fun checkPayloadJoinsSixFields() {
        assertEquals(
            "1\n1.1.0\nroot\nu\n10\nabc",
            CheckPayload.build(1, "1.1.0", "root", "u", 10, "abc"),
        )
        assertEquals(
            "\n\n\n\n\n",
            CheckPayload.build(null, null, null, null, null, null),
        )
    }

    @Test
    fun defaultUpdaterAdvertisesZipButNotDeltaOrFileList() {
        val transport = RecordingTransport { jsonOk(CHECK_200) }
        val client = Client(ClientConfig("http://example.test", "p", transport = transport))
        val (caps, algos) = Updater(client).capabilities()
        assertEquals(listOf(Capability.FULL_PACKAGE, Capability.PATCH_PACKAGE), caps)
        assertEquals(null, algos)
    }

    @Test
    fun packReadyDoesNotCompareArchiveHashToFullPackageSha() {
        val hasher = JdkHasher()
        val payload = "file-a".toByteArray()
        val fileHash = hasher.sha256Hex(payload)
        val zipBytes = java.io.ByteArrayOutputStream().use { out ->
            ZipOutputStream(out).use { zip ->
                zip.putNextEntry(ZipEntry(fileHash))
                zip.write(payload)
                zip.closeEntry()
            }
            out.toByteArray()
        }
        val zipHash = hasher.sha256Hex(zipBytes)
        val check = CHECK_200
            .replace("\"package_type\": \"single_file\"", "\"package_type\": \"multi_file\"")
            .replace("\"sha256\": \"aaaa\"", "\"sha256\": \"ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff\"")
        val integrity = """
            {
              "version_semver": "1.1.0",
              "channel": "stable",
              "package_type": "multi_file",
              "root_hash": "",
              "full_package_url": "/full",
              "file_name": "pkg.zip",
              "size": 1,
              "sha256": "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff",
              "files": [{"path":"bin/app","size":${payload.size},"install_policy":"OVERWRITE","integrity_check":true,"sha256":"$fileHash"}]
            }
        """.trimIndent()
        val transport = RecordingTransport { req ->
            when {
                RecordingTransport.pathOf(req.url).endsWith("/update/check") -> jsonOk(check)
                RecordingTransport.pathOf(req.url).contains("/integrity") -> jsonOk(integrity)
                RecordingTransport.pathOf(req.url).endsWith("/update/pack") -> jsonOk(
                    """{"status":"ready","package_url":"/pkg.zip","sha256":"$zipHash","files":[{"path":"bin/app","size":${payload.size},"install_policy":"OVERWRITE","integrity_check":true,"sha256":"$fileHash"}]}""",
                )
                RecordingTransport.pathOf(req.url).endsWith("/pkg.zip") -> HttpResponse(200, body = zipBytes)
                RecordingTransport.pathOf(req.url).endsWith("/telemetry/report") -> jsonOk("{}", 202)
                else -> errorJson(404, "NOT_FOUND", RecordingTransport.pathOf(req.url))
            }
        }
        val client = Client(ClientConfig("http://example.test", "demo", transport = transport))
        val tmp = Files.createTempDirectory("kv-pack-ready")
        val store = JdkFileStore(tmp)
        val outcome = Updater(client, fileStore = store).run(
            UpdateIdentity("1.0.0", "windows", "x86_64"),
        )
        assertEquals("patch_package", outcome.diffMode)
        assertFalse(outcome.fallbackToFull)
        assertTrue(payload.contentEquals(store.read("bin/app")))
        assertFalse(store.exists("package.bin"))
    }

    @Test
    fun replacerReceivesFileStorePathNotInstallRootRelative() {
        val payload = byteArrayOf(1, 2, 3, 4)
        val hash = JdkHasher().sha256Hex(payload)
        val check = CHECK_200.replace("\"sha256\": \"aaaa\"", "\"sha256\": \"$hash\"")
            .replace("\"size\": 4", "\"size\": ${payload.size}")
        val transport = RecordingTransport { req ->
            when {
                RecordingTransport.pathOf(req.url).endsWith("/update/check") -> jsonOk(check)
                RecordingTransport.pathOf(req.url).endsWith("/telemetry/report") -> jsonOk("{}", 202)
                else -> HttpResponse(200, body = payload)
            }
        }
        val client = Client(ClientConfig("http://example.test", "demo", transport = transport))
        val stage = Files.createTempDirectory("kv-stage")
        val install = Files.createTempDirectory("kv-install")
        var seenStaged: java.nio.file.Path? = null
        var seenInstall: java.nio.file.Path? = null
        val replacer = Replacer { staged, current ->
            seenStaged = staged
            seenInstall = current
        }
        val outcome = Updater(
            client,
            fileStore = JdkFileStore(stage),
            archiveUnpacker = null,
            replacer = replacer,
        ).run(
            UpdateIdentity("1.0.0", "windows", "x86_64", installRoot = install),
        )
        assertEquals(UpdateKind.APPLIED, outcome.kind)
        assertTrue(outcome.applied)
        assertEquals(stage.resolve("package.bin").toAbsolutePath().normalize(), seenStaged?.toAbsolutePath()?.normalize())
        assertEquals(install.toAbsolutePath().normalize(), seenInstall?.toAbsolutePath()?.normalize())
        assertFalse(install.resolve("package.bin").toFile().exists())
    }

    @Test
    fun updaterForwardsCustomOnDeviceReport() {
        val transport = RecordingTransport { req ->
            when {
                RecordingTransport.pathOf(req.url).endsWith("/clients/report") ->
                    jsonOk("""{"ip":"127.0.0.1","country_code":"","region_code":"","geo_i18n":{}}""")
                RecordingTransport.pathOf(req.url).endsWith("/update/check") -> jsonOk("", 204)
                else -> errorJson(404, "NOT_FOUND", RecordingTransport.pathOf(req.url))
            }
        }
        val client = Client(ClientConfig("http://example.test", "p", transport = transport))
        Updater(client, archiveUnpacker = null).run(
            UpdateIdentity(
                currentVersion = "1.0.0",
                os = "windows",
                arch = "x86_64",
                deviceId = "dev-custom",
                custom = JsonObject(mapOf("k" to JsonPrimitive("v"))),
                reportDevice = true,
            ),
        )
        val report = transport.requests.first { RecordingTransport.pathOf(it.url).endsWith("/clients/report") }
        val body = KiriversJson.parseToJsonElement(report.body!!.decodeToString()).jsonObject
        assertEquals("v", body["custom"]!!.jsonObject["k"].toString().trim('"'))
    }
}

File: tests/adapter_test.cpp
#include "require.hpp"

#include "delta.hpp"
#include <kirivers/kirivers.hpp>

#include <nlohmann/json.hpp>
#include <chrono>
#include <map>
#include <string>
#include <vector>

using namespace kirivers;

int main() {
  {
    auto a = normalize_rel_path(R"(bin\resources\config.json)");
    REQUIRE(a.has_value());
    REQUIRE(*a == "bin/resources/config.json");
    auto b = normalize_rel_path("bin//sub\\\\dir///file.txt");
    REQUIRE(b.has_value());
    REQUIRE(*b == "bin/sub/dir/file.txt");
    REQUIRE(!normalize_rel_path("../etc/passwd"));
    REQUIRE(!normalize_rel_path("/abs"));
    REQUIRE(!normalize_rel_path("C:foo"));
    REQUIRE(equal_hex("AaBb", "aabb"));
    REQUIRE(!equal_hex("aa", "ab"));
  }

#if defined(KIRIVERS_HOSTED)
  {
    const std::string nfd = "caf\u0065\u0301.txt";
    const std::string nfc = "caf\u00e9.txt";
    auto a = normalize_rel_path(nfd, utf8proc_nfc);
    auto b = normalize_rel_path(nfc, utf8proc_nfc);
    REQUIRE(a.has_value());
    REQUIRE(b.has_value());
    REQUIRE(*a == *b);
  }
#endif

  {
    Bytes hd{'H', 'D', 'I', 'F', 'F', '1', '3', '&'};
    REQUIRE(detect_delta_magic(hd) == DeltaMagic::Hdiff13);
    REQUIRE(magic_matches_algo(DeltaMagic::Hdiff13, "hdiffpatch"));
    REQUIRE(!magic_matches_algo(DeltaMagic::Hdiff13, "xdelta3"));
    Bytes kv{'K', 'V', 'D', 'I', 'F', 'F', 'H', 'P', '1', '\n'};
    REQUIRE(detect_delta_magic(kv) == DeltaMagic::KvDiffHp1);
    REQUIRE(magic_matches_algo(DeltaMagic::KvDiffHp1, "hdiffpatch"));
    REQUIRE(!magic_matches_algo(DeltaMagic::KvDiffHp1, "bsdiff"));
    Bytes bs{'B', 'S', 'D', 'I', 'F', 'F', '4', '0'};
    REQUIRE(detect_delta_magic(bs) == DeltaMagic::Bsdiff40);
    Bytes vc{static_cast<unsigned char>(0xD6), static_cast<unsigned char>(0xC3),
             static_cast<unsigned char>(0xC4)};
    REQUIRE(detect_delta_magic(vc) == DeltaMagic::Vcdiff);
    Bytes unk{'X'};
    REQUIRE(detect_delta_magic(unk) == DeltaMagic::Unknown);
  }

  std::vector<HttpRequest> calls;
  int pack_hits = 0;
  std::string first_pack_body;
  auto handler = [&](const HttpRequest& req) -> HttpResponse {
    calls.push_back(req);
    HttpResponse r;
    r.status = 200;
    if (req.url.find("/update/check") != std::string::npos) {
      auto j = nlohmann::json::parse(bytes_to_string(req.body));
      r.headers["Content-Type"] = "application/json";
      auto body = j.dump();
      r.body.assign(body.begin(), body.end());
      r.status = 204;
      return r;
    }
    if (req.url.find("/update/pack") != std::string::npos) {
      ++pack_hits;
      const std::string b = bytes_to_string(req.body);
      if (pack_hits == 1) first_pack_body = b;
      REQUIRE(b == first_pack_body);
      nlohmann::json out;
      if (pack_hits < 2) {
        r.status = 202;
        out = {{"status", "pending"}};
      } else {
        r.status = 200;
        out = {{"status", "ready"},
               {"package_url", "/p"},
               {"diff_mode", "patch_package"}};
      }
      auto s = out.dump();
      r.body.assign(s.begin(), s.end());
      return r;
    }
    return r;
  };

  {
    Config cfg;
    cfg.base_url = "http://example.test";
    cfg.project_ref = "p";
    cfg.hosted_defaults = false;
    cfg.transport = std::make_shared<FunctionTransport>(handler);
    Client client(cfg);
    auto caps = client.declared_capabilities();
    REQUIRE(caps.capabilities.size() == 1);
    REQUIRE(caps.capabilities[0] == "full_package");
    REQUIRE(caps.accepted_delta_algos.empty());

    CheckInput in;
    in.current_version = "1.0.0";
    in.os = "linux";
    in.arch = "x86_64";
    client.check(in);
    auto j = nlohmann::json::parse(bytes_to_string(calls.back().body));
    REQUIRE(j["capabilities"] == nlohmann::json::array({"full_package"}));
    REQUIRE(!j.contains("accepted_delta_algos"));

    CheckInput empty_caps = in;
    empty_caps.capabilities = std::vector<std::string>{};
    client.check(empty_caps);
    auto j_empty = nlohmann::json::parse(bytes_to_string(calls.back().body));
    REQUIRE(j_empty["capabilities"] == nlohmann::json::array({"full_package"}));
  }

  {
    calls.clear();
    Config cfg;
    cfg.base_url = "http://example.test";
    cfg.project_ref = "p";
    cfg.hosted_defaults = false;
    cfg.transport = std::make_shared<FunctionTransport>(handler);
    cfg.patcher = std::make_shared<FunctionPatcher>(
        std::vector<std::string>{"bsdiff", "xdelta3", "hdiffpatch"},
        [](const Bytes&, const Bytes& d, const std::string&) { return d; });
    class DummyUnpack final : public ArchiveUnpacker {
     public:
      void unpack_zip(const Bytes&, FileStore&, const std::string&,
                      const std::vector<std::pair<std::string, std::string>>&)
          override {}
    };
    class DummyStore final : public FileStore {
     public:
      Bytes read_file(const std::string&) const override { return {}; }
      void write_file(const std::string&, const Bytes&) override {}
      bool exists(const std::string&) const override { return false; }
      std::vector<std::string> list_files(const std::string&) const override {
        return {};
      }
      bool can_write_individual_files() const override { return true; }
    };
    cfg.archive_unpacker = std::make_shared<DummyUnpack>();
    cfg.file_store = std::make_shared<DummyStore>();
    Client client(cfg);
    auto caps = client.declared_capabilities();
    bool has_delta = false, has_patch = false, has_files = false;
    for (const auto& c : caps.capabilities) {
      if (c == "binary_delta") has_delta = true;
      if (c == "patch_package") has_patch = true;
      if (c == "file_list") has_files = true;
    }
    REQUIRE(has_delta);
    REQUIRE(has_patch);
    REQUIRE(has_files);
    REQUIRE(!caps.accepted_delta_algos.empty());

    CheckInput in;
    in.current_version = "1.0.0";
    in.os = "linux";
    in.arch = "x86_64";
    client.check(in);
    auto j = nlohmann::json::parse(bytes_to_string(calls.back().body));
    REQUIRE(j.contains("accepted_delta_algos"));
    REQUIRE(j["accepted_delta_algos"].size() == 3);
  }

  {
    pack_hits = 0;
    first_pack_body.clear();
    Config cfg;
    cfg.base_url = "http://example.test";
    cfg.project_ref = "p";
    cfg.hosted_defaults = false;
    cfg.transport = std::make_shared<FunctionTransport>(handler);
    Client client(cfg);
    PackInput pin;
    pin.source_version = "1.0.0";
    pin.target_version = "1.1.0";
    pin.os = "linux";
    pin.arch = "x86_64";
    pin.needed_paths = {"a/b", "a\\b"};
    PackWaitOptions opt;
    opt.initial = std::chrono::milliseconds(1);
    opt.cap = std::chrono::milliseconds(1);
    opt.deadline = std::chrono::milliseconds(2000);
    auto packed = client.pack_wait(pin, opt);
    REQUIRE(packed.status == "ready");
    REQUIRE(pack_hits == 2);
  }

  {
    const std::string sha(64, 'a');
    class MemStore final : public FileStore {
     public:
      std::map<std::string, Bytes> files;
      Bytes read_file(const std::string& p) const override {
        auto it = files.find(p);
        if (it == files.end()) throw std::runtime_error("missing");
        return it->second;
      }
      void write_file(const std::string& p, const Bytes& d) override { files[p] = d; }
      bool exists(const std::string& p) const override { return files.count(p) != 0; }
      std::vector<std::string> list_files(const std::string&) const override { return {}; }
    };
    auto store = std::make_shared<MemStore>();
    int patch_apply = 0;
    auto h = [&](const HttpRequest& req) -> HttpResponse {
      HttpResponse r;
      r.status = 200;
      auto json_body = [&](const nlohmann::json& j) {
        r.headers["Content-Type"] = "application/json";
        auto s = j.dump();
        r.body.assign(s.begin(), s.end());
        return r;
      };
      if (req.url.find("/update/check") != std::string::npos) {
        return json_body({{"has_update", true},
                          {"is_downgrade", false},
                          {"compare_engine", "semver"},
                          {"version_semver", "1.1.0"},
                          {"target_channel", "stable"},
                          {"package_type", "single_file"},
                          {"root_hash", ""},
                          {"package_url", "http://example.test/full.bin"},
                          {"file_name", "app.bin"},
                          {"size", 4},
                          {"sha256", sha},
                          {"delta_available", true},
                          {"delta_algo", "bsdiff"}});
      }
      if (req.url.find("/update/diff") != std::string::npos) {
        return json_body({{"diff_mode", "binary_delta"},
                          {"root_hash", ""},
                          {"version_semver", "1.1.0"},
                          {"channel", "stable"},
                          {"compare_engine", "semver"},
                          {"package_url", "http://example.test/delta.bin"},
                          {"sha256", sha},
                          {"delta_algo", "bsdiff"}});
      }
      if (req.url.find("/delta.bin") != std::string::npos) {
        r.body = {'X', 'X', 'X'};
        return r;
      }
      if (req.url.find("/full.bin") != std::string::npos) {
        r.body = {'F', 'U', 'L', 'L'};
        return r;
      }
      if (req.url.find("/telemetry/report") != std::string::npos) {
        r.status = 202;
        return r;
      }
      return r;
    };
    Config cfg;
    cfg.base_url = "http://example.test";
    cfg.project_ref = "p";
    cfg.hosted_defaults = false;
    cfg.transport = std::make_shared<FunctionTransport>(h);
    cfg.file_store = store;
    cfg.hasher = std::make_shared<FunctionHasher>([&](const Bytes&) { return sha; });
    cfg.patcher = std::make_shared<FunctionPatcher>(
        std::vector<std::string>{"bsdiff"},
        [&](const Bytes&, const Bytes& d, const std::string&) {
          ++patch_apply;
          return d;
        });
    Client client(cfg);
    Updater updater(client);
    UpdateRequest ureq;
    ureq.current_version = "1.0.0";
    ureq.os = "linux";
    ureq.arch = "x86_64";
    ureq.stage_path = "staged.bin";
    auto result = updater.run(ureq);
    REQUIRE(result.kind == UpdateKind::Staged);
    REQUIRE(result.diff_mode == "full_package");
    REQUIRE(patch_apply == 0);
    REQUIRE(store->files["staged.bin"] == Bytes({'F', 'U', 'L', 'L'}));
  }

  {
    const std::string sha(64, 'b');
    class MemStore final : public FileStore {
     public:
      std::map<std::string, Bytes> files;
      Bytes read_file(const std::string& p) const override {
        auto it = files.find(p);
        if (it == files.end()) throw std::runtime_error("missing");
        return it->second;
      }
      void write_file(const std::string& p, const Bytes& d) override { files[p] = d; }
      bool exists(const std::string& p) const override { return files.count(p) != 0; }
      std::vector<std::string> list_files(const std::string&) const override { return {}; }
    };
    class RecUnpack final : public ArchiveUnpacker {
     public:
      int calls = 0;
      std::string dest;
      void unpack_zip(const Bytes&, FileStore&, const std::string& dest_root,
                      const std::vector<std::pair<std::string, std::string>>& m) override {
        ++calls;
        dest = dest_root;
        REQUIRE(!m.empty());
        REQUIRE(m.front().second == "bin/app");
      }
    };
    auto store = std::make_shared<MemStore>();
    auto unpack = std::make_shared<RecUnpack>();
    auto h = [&](const HttpRequest& req) -> HttpResponse {
      HttpResponse r;
      r.status = 200;
      auto json_body = [&](const nlohmann::json& j) {
        r.headers["Content-Type"] = "application/json";
        auto s = j.dump();
        r.body.assign(s.begin(), s.end());
        return r;
      };
      if (req.url.find("/update/check") != std::string::npos) {
        return json_body({{"has_update", true},
                          {"is_downgrade", false},
                          {"compare_engine", "semver"},
                          {"version_semver", "1.1.0"},
                          {"target_channel", "stable"},
                          {"package_type", "multi_file"},
                          {"root_hash", ""},
                          {"package_url", "http://example.test/full.zip"},
                          {"file_name", "app.zip"},
                          {"size", 1},
                          {"sha256", sha},
                          {"delta_available", false}});
      }
      if (req.url.find("/integrity") != std::string::npos) {
        return json_body(
            {{"version_semver", "1.1.0"},
             {"channel", "stable"},
             {"package_type", "multi_file"},
             {"root_hash", ""},
             {"full_package_url", "http://example.test/full.zip"},
             {"file_name", "app.zip"},
             {"size", 1},
             {"sha256", sha},
             {"files", nlohmann::json::array(
                           {{{"path", "bin/app"},
                             {"size", 1},
                             {"install_policy", "REPLACE"},
                             {"integrity_check", true},
                             {"sha256", sha}}})}});
      }
      if (req.url.find("/update/pack") != std::string::npos) {
        return json_body({{"status", "ready"},
                          {"diff_mode", "patch_package"},
                          {"package_url", "http://example.test/patch.zip"},
                          {"sha256", sha},
                          {"files", nlohmann::json::array(
                                        {{{"path", "bin/app"}, {"sha256", sha}}})}});
      }
      if (req.url.find("/patch.zip") != std::string::npos) {
        r.body = {'Z', 'I', 'P'};
        return r;
      }
      if (req.url.find("/telemetry/report") != std::string::npos) {
        r.status = 202;
        return r;
      }
      return r;
    };
    Config cfg;
    cfg.base_url = "http://example.test";
    cfg.project_ref = "p";
    cfg.hosted_defaults = false;
    cfg.transport = std::make_shared<FunctionTransport>(h);
    cfg.file_store = store;
    cfg.archive_unpacker = unpack;
    cfg.hasher = std::make_shared<FunctionHasher>([&](const Bytes&) { return sha; });
    Client client(cfg);
    Updater updater(client);
    UpdateRequest ureq;
    ureq.current_version = "1.0.0";
    ureq.os = "linux";
    ureq.arch = "x86_64";
    ureq.install_dir = "app";
    ureq.stage_path = "staged.zip";
    auto result = updater.run(ureq);
    REQUIRE(result.kind == UpdateKind::Staged);
    REQUIRE(result.diff_mode == "patch_package");
    REQUIRE(unpack->calls == 1);
    REQUIRE(unpack->dest == "app");
    REQUIRE(store->files["staged.zip"] == Bytes({'Z', 'I', 'P'}));
  }

#if defined(KIRIVERS_HOSTED)
  {
    calls.clear();
    Config cfg;
    cfg.base_url = "http://example.test";
    cfg.project_ref = "p";
    cfg.hosted_defaults = true;
    cfg.transport = std::make_shared<FunctionTransport>(handler);
    Client client(cfg);
    auto caps = client.declared_capabilities();
    bool has_full = false, has_delta = false, has_patch = false, has_files = false;
    for (const auto& c : caps.capabilities) {
      if (c == "full_package") has_full = true;
      if (c == "binary_delta") has_delta = true;
      if (c == "patch_package") has_patch = true;
      if (c == "file_list") has_files = true;
    }
    REQUIRE(has_full);
    REQUIRE(has_patch);
    REQUIRE(has_files);
    REQUIRE(!has_delta);
    REQUIRE(caps.accepted_delta_algos.empty());
    REQUIRE(client.replacer() == nullptr);
  }
#endif

#if defined(KIRIVERS_EMBEDDED)
  {
    Config cfg;
    cfg.base_url = "http://example.test";
    cfg.project_ref = "p";
    bool threw = false;
    try {
      Client client(cfg);
      (void)client;
    } catch (const ConfigError&) {
      threw = true;
    }
    REQUIRE(threw);
  }
#endif

  std::cout << "adapter_test ok\n";
  return 0;
}
